File: 프로그래머스/4/43236. 징검다리/징검다리.py
def solution(distance, rocks, n):
    rocks.append(distance)
    rocks.sort()
    
    answer = 0
    left = 0
    right = distance + 1
    
    while left < right:
        mid = (left + right) // 2
        
        prev = 0
        count = 0
        
        for _, r in enumerate(rocks):
            if r - prev < mid:
                count += 1
            else:
                prev = r
            
            if count > n:
                break
        
        if count > n:
            right = mid
        else:
            answer = mid
            left = mid + 1
    
    return answer

File: 프로그래머스/4/43236. 징검다리/징검다리.java
import java.util.Arrays;

class Solution {

    public int solution(int distance, int[] rocks, int n) {
        Arrays.sort(rocks);

        int low = 1;
        int high = distance;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canMaintainDistance(rocks, distance, n, mid)) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return high;
    }

    private boolean canMaintainDistance(
            int[] rocks,
            int distance,
            int maxRemovals,
            int minDistance
    ) {
        int removed = 0;
        int previous = 0;

        for (int rock : rocks) {
            if (rock - previous < minDistance) {
                removed++;

                if (removed > maxRemovals) {
                    return false;
                }
            } else {
                previous = rock;
            }
        }
        
        if (distance - previous < minDistance) {
            removed++;
        }

        return removed <= maxRemovals;
    }
}